[FEAT] API E2E 테스트 추가

dye0p opened the pull request:
## Issue Number
closed #91

## As-Is
<!-- 문제 상황 정의 -->
- API E2E 테스트의 부재로 다양한 시나리오에 대한 테스트가 없던 상황

## To-Be
<!-- 변경 사항 -->
- API E2E 테스트를 추가하여, 다양한 시나리오의 테스트를 구현
- 정상적인 요청과 응답, 정상적이지 않은 요청, 응답을 테스트

## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?

## (Optional) Additional Description
- API 테스트를 위해 `RestAssured`를 사용하였습니다.

### CategoryController API 테스트 목록
- 카테고리 목록 조회
- 등록된 카테고리가 없는 경우

### MentoringController API 테스트 목록
- 필터링 없이 모든 멘토링 목록 조회
- 저장된 멘토링이 없는 경우 목록 조회
- 특정 카테고리를 모두 포함하는 멘토링 조회 (AND 조건 필터링)
- 필터 조건에 해당하는 멘토링이 없는 경우
- 존재하지 않는 카테고리로 필터링 시도
- 존재하는 멘토링 ID로 조회
- 존재하지 않는 멘토링 ID로 조회 시도

### ReservationController API 테스트 목록
외부 API(`SmsRestClientService`)는 `@MockitoBean`을 사용해 실제 호출 없이 동작하도록 처리
- 유효한 예약 요청 시
- 잘못된 예약 정보(예: 잘못된 휴대폰 번호)
- 존재하지 않는 멘토링에 예약 요청

### `@Disabled` 테스트 안내
전역 예외 처리 미완성 / 구현 대기 / 명확한 예외 메시지 정의 미흡 등의 이유로 아래 테스트들은 현재 `@Disabled` 처리되어 있습니다. 예외 처리 로직 반영 후 활성화 예정입니다.

dye0p committed:
test: CategoryController API 테스트 추가

- 모든 카테고리 목록 조회 e2e 테스트 추가
- 빈 카테고리 목록 조회 e2e2 테스트 추가
Co-authored-by: jye326 <juyongeun@gmail.com>

dye0p committed:
test: MentoringController API 테스트 추가

- 멘토링 전체 조회 e2e 테스트 추가
- 멘토링 필터링 조회 e2e2 테스트 추가

dye0p committed:
test: ReservationController API 테스트 추가

- 멘토링 예약 e2e 테스트 추가
- Sms API MockBean 처리

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Category.java:
```diff
@@ -25,6 +27,6 @@ public class Category {
 
     public Category(String title) {
-        this.title = title;
+        this(null, title);
```
컨벤션 지킴이 👍

jye326 commented on backend/fittoring/src/test/java/fittoring/mentoring/presentation/api/CategoryControllerTest.java:
```diff
@@ -0,0 +1,90 @@
+        dbCleaner.clean();
+    }
+
+    @DisplayName("카테고리 목록 조회가 성공하면, 200 OK 상태고리와 카테고리 목록을 반환한다.")
```
p2 : 카테고리에 너무 집중하다가 상태코드가 상태고리가 되어버린 듯 하네요

jye326 commented on backend/fittoring/src/test/java/fittoring/mentoring/presentation/api/CategoryControllerTest.java:
```diff
@@ -0,0 +1,90 @@
+
+        assertThat(response)
+                .hasSize(3)
+                .contains(expected, expected2, expected3);
```
p3 : containsExactly를 사용하여 hasSize와 contains를 한번에 검사할 수 있을 것 같아요. 만약 적용한다면 다른 전체 조회 테스트들에도 일괄적용해야겠구요.

dye0p replied:
JPA의 findAll 메서드는 별다른 정렬 조건을 명시해주지 않으면 순서 보장이 되지 않습니다. 그래서 순서 검증까지 할 수 없어서 contains를 사용하여 검증하였습니다!

jye326 commented on backend/fittoring/src/test/java/fittoring/mentoring/presentation/api/CategoryControllerTest.java:
```diff
@@ -0,0 +1,90 @@
+                .contains(expected, expected2, expected3);
+    }
+
+    @DisplayName("등록된 카테고리가 없을 때, 200 OK 상태고리와 빈 카테고리 목록을 반환한다.")
```
여기도 상태고리😀

jye326 commented on backend/fittoring/src/test/java/fittoring/mentoring/presentation/api/ReservationControllerTest.java:
```diff
@@ -0,0 +1,207 @@
+        assertThat(response.statusCode()).isEqualTo(400);
+    }
+
+    // 전역 예외처리 도입 후 테스트 활성화 필요(@Disabled 어노테이션 제거)
```
p1 : 이거 해 주세요!

jye326 reviewed (commented):
꼼꼼한 테스트 수고하셨습니다~ 코멘트 반영 후 다시 요청해주세요!

kimdevv commented on backend/fittoring/src/test/java/fittoring/mentoring/presentation/api/CategoryControllerTest.java:
```diff
@@ -0,0 +1,90 @@
+        //when
+        List<CategoryResponse> response = RestAssured
+                .given()
+                .log().all().contentType(ContentType.JSON)
```
p3: `.log().all()`을 작성해주신 이유가 궁금합니다!

dye0p replied:
해당 메서드를 통해서 HTTP 요청정보를 console에 출력하기 위해서 사용하였습니다!

kimdevv commented on backend/fittoring/src/test/java/fittoring/mentoring/presentation/api/CategoryControllerTest.java:
```diff
@@ -0,0 +1,90 @@
+
+        assertThat(response)
+                .hasSize(3)
+                .contains(expected, expected2, expected3);
```
p3: .`hasSize().contains()`는 `.containsExactlyInAnyOrder()`으로 단축할 수 있을 것 같습니다! 해당 메서드는 **순서를 제외하고 개수와 요소가 일치하는지** 검증해줍니다!

dye0p replied:
좋네요! 감사합니다 : )

kimdevv commented on backend/fittoring/src/test/java/fittoring/mentoring/presentation/api/ReservationControllerTest.java:
```diff
@@ -0,0 +1,207 @@
+    @Test
+    void createReservation() {
+        //given
+        doNothing()
```
모키토에서 `doNothing()`이라는 기능도 제공하고 있군요!

kimdevv reviewed (approved):
고생하셨습니다~

dye0p committed:
test: 테스트 디스플레이 네이밍 오타 수정

dye0p committed:
test: 테스트 검증 메서드 수정

- containsExactlyInAnyOrder 메서드를 통해, 개수와, 내용을 모두 검증할 수 있다.

jye326 reviewed (approved):
인정합니다

ppparkta commented on backend/fittoring/src/test/java/fittoring/mentoring/presentation/api/CategoryControllerTest.java:
```diff
@@ -0,0 +1,88 @@
+                .then().log().all()
+                .statusCode(200)
+                .extract()
+                .as(new TypeRef<>() {
```
p3: `TypeRef` 이것은 왜 쓰셨나요?

dye0p replied:
반환되는 List의 제네릭 타입을 명시해주기 위해서 사용했습니다.
사용하지 않는다면 `.as(List.class)`로 명시해야 하는데, 이는 정확한 반환 타입을 알 수 없기 때문입니다!

ppparkta commented on backend/fittoring/src/test/java/fittoring/mentoring/presentation/api/CategoryControllerTest.java:
```diff
@@ -0,0 +1,88 @@
+        CategoryResponse expected2 = new CategoryResponse(savedCategory2.getId(), savedCategory2.getTitle());
+        CategoryResponse expected3 = new CategoryResponse(savedCategory3.getId(), savedCategory3.getTitle());
+
+        assertThat(response).containsExactlyInAnyOrder(expected, expected2, expected3);
```
👍

ppparkta commented on backend/fittoring/src/test/java/fittoring/mentoring/presentation/api/CategoryControllerTest.java:
p2: 통합테스트 성격이 강한 것 같습니다. 해당 테스트는 통합테스트 패키지에 있는 것이 적절하지 않을까요? 다른 컨트롤러 테스트 모두 포함해서요!

dye0p replied:
동의합니다! 모두 integration 패키지로 이동하였습니다.

ppparkta commented on backend/fittoring/src/test/java/fittoring/mentoring/presentation/api/ReservationControllerTest.java:
```diff
@@ -0,0 +1,207 @@
+    private SmsRestClientService smsRestClientService;
+
+    @Autowired
+    private DbCleaner dbCleaner;
```
p3: 해당 필드를 하단에 배치하는 것으로 통일하는 것은 어떠신가요?

dye0p replied:
좋습니다👍

ppparkta commented on backend/fittoring/src/test/java/fittoring/mentoring/presentation/api/ReservationControllerTest.java:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    // 전역 예외처리 도입 후 테스트 활성화 필요(@Disabled 어노테이션 제거)
+    @Disabled
```
p1: 제거 안 된 것 같아요!

dye0p replied:
해당 이슈 작업 기준 예외처리가 되지 않아서 Disabled 했었는데, 현재는 저장소의 최신 코드를 받아온 상태여서 @Disabled를 제거했습니다!

ppparkta reviewed (commented):
테스트 많아서 시간이 오래 걸렸을 것 같은데 너무 잘 해주셨네요!! 수고하셨어요😄

dye0p committed:
Merge branch 'develop-be' into feature/91-e2e-test
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

dye0p committed:
test: 존재하지 않는 멘토링에 대한 예약 테스트, 잘못된 부분 수정 후 @Disabled 제거

- 존재하지 않는 멘토링에 대한 예약을 검증하는 테스트지만, 잘못된 요청(잘못된 휴대폰 번호)이 들어가고 있었던 문제로 테스트가 의도한대로 동작하지 않았음.
- 정상적인 요청을 할 수 있도록 수정하였음

dye0p committed:
test: 존재하지 않는 멘토링 Id 조회 테스트 @Disabled 제거

- 작업 이슈에서 전역 예외처리 부분이 존재하지 않아 Disabled 했지만, 현재 커밋 시점에서 최신 사항을 받아왔기 때문에 Disabled 제거

dye0p committed:
test: DbClear.class 필드 위치 하단으로 이동

- 다른 테스트와 동일한 위치로 통일시키기 위해서 이동

dye0p committed:
test: 테스트 패키지 이동(controller -> integration)

- e2e 테스트는 통합테스트에 포함되기 때문에, controller 패키지에 존재하는것이 어색하다고 판단하여 integration 패키지로 이동

dye0p reviewed (commented):
수정 완료했습니다!
